en/tensor-examples.html

```diff
@@ -0,0 +1,246 @@
+
+<pre>
+concat(myDenseTensor(), tensor(x[3]):[myAttribute(1), myAttribute(2), myScalarFunction()], x)
+</pre>
+
+<p>This creates a tensor from a set of scalar expressions, and concatenates it to the dense tensor.</p>
+
+
+<h2 id="dot-product-between-query-and-document-vectors">Dot Product between query and document vectors</h2>
+
+<p>Assume we have a set of documents where each document contains a vector of size 4.
+We want to calculate the dot product between the document vectors and a vector passed down with the query
+and rank the results according to the dot product score.</p>
+
+<p>The following schema file defines an attribute tensor field
+with a tensor type that has one indexed dimension <code>x</code> of size 4.
+In addition we define a rank profile that calculates the dot product.</p>
+
+<pre>
+schema example {
+  document example {
+    field document_vector type tensor&lt;float&gt;(x[4]) {
+      indexing: attribute | summary
+    }
+  }
+  rank-profile dot_product {
+    first-phase {
+      expression: sum(query(query_vector)*attribute(document_vector))
+    }
+  }
+}
+</pre>
+
+<p>The tensor to pass down with query is defined in a query profile type with the same tensor type as
+the field in the document:</p>
+
+<pre>
+&lt;query-profile-type id="myProfileType"&gt;
+  &lt;field name="ranking.features.query(query_vector)" type="tensor&amp;lt;float&amp;gt;(x[4])" /&gt;
+&lt;/query-profile-type&gt;
+</pre>
+
+<p>Example document with the vector [1.0, 2.0, 3.0, 5.0]:</p>
+
+<pre>
+[
+  { "put": "id:example:example::0", "fields": {
+      "document_vector" : {
+        "cells": [
+          { "address" : { "x" : "0" }, "value": 1.0 },
+          { "address" : { "x" : "1" }, "value": 2.0 },
+          { "address" : { "x" : "2" }, "value": 3.0 },
+          { "address" : { "x" : "3" }, "value": 5.0 }
+        ]
+      }
+    }
+  }
+]
+</pre>
+
+<p>Example query set in a searcher with the vector [1.0, 2.0, 3.0, 5.0]:</p>
+
+<pre>
+public Result search(Query query, Execution execution) {
+    query.getRanking().getFeatures().put("query(query_vector)",
+        Tensor.Builder.of(TensorType.fromSpec("tensor&lt;float&gt;(x[4])")).
+        cell().label("x", 0).value(1.0).
+        cell().label("x", 1).value(2.0).
+        cell().label("x", 2).value(3.0).
+        cell().label("x", 3).value(5.0).build());
+    return execution.search(query);
+}
+</pre>
+
+
+<h2>Logistic regression models with cross features</h2>
+
+<p>One simple way to use machine-learning is to generate cross features from a set of base features
+and then do a logistic regression on these. How can this be expressed as Vespa tensors?</p>
+
+<p>Assume we have three base features:</p>
+
+<pre>
+query(interests): tensor(interest{}) - A sparse, weighted set of the interests of a user.
+query(location): tensor(location{}) - A sparse set of the location(s) of the user.
+attribute(topics): tensor(topic{}) - A sparse, weighted set of the topics of a given document.
+</pre>
```

Comment:
these should be called mapped tensors

Reply:
Here I call them sets. A tensor dimension is mapped or indexed. A tensor can be a map, set, matrix, vector etc, and can be sparse, dense, mixed etc.